Option to current scale with VREF on TMC2208/9

mihtjel opened the pull request:
### Current-scaling using VREF for TMC2208 and TMC2209
This allows fine-grained control of current for boards whose designs allow high current, but where smaller steppers are used. It also allows limiting the maximum current in case of misconfiguration of IRUN/IHOLD, and for external analog scaling of the current directly at the driver IC.

### Description

The stepper drivers TMC2208 and TMC2209 are currently reasonably well-supported in Marlin using UART control. They do, however, offer many more options than are currently made available to users.  This pull-request is the first of maybe a couple that will attempt to address some of these options.

At hand is the option to use the VREF input to scale the set current for each driver. A common TMC2209 based board might have sense resistors of 100mOhm, allowing a maximum peak current of 2.7A, and an RMS current of 1.92A. This gives a step resolution of 59.8mA RMS per step, or ~85mA peak. 

A user wishing to drive a 1A limited standard NEMA17 40mm stepper at roughly 80% current, or 800mA, could do so comfortably by setting VREF at 1.25V, putting a current limit of 1.35A peak / 0.95A RMS on the driver, and increasing the resolution of the current steps to 29.9mA.

Fine tuning by external analogue circuits would also be possible, as would be simple safety devices such as a temperature sensitive voltage divider limiting the current before the chip overheats.

(Enough of my rambling: I just thought it'd be cool.)

### Requirements

Requires TMC2208/9 stepper drivers with UART control, and some form of adjustable
VREF input.

### Benefits

Adds fine-grained current control beyond the existing steps for TMC drivers.

### Configurations

(Might do this later, but for now I'm gonna go to bed.)

### Related Issues

No, just my own idea.

mihtjel committed:
Option to current scale with VREF on TMC2208/9
This allows fine-grained control of current for boards whose designs allow
 high current, but where smaller steppers are used. It also allows limiting
 the maximum current in case of misconfiguration of IRUN/IHOLD, and
 for external analog scaling of the current directly at the driver IC.

mihtjel commented:
Looks like it might also work for TMC2130, 5130 and 5160, but I don't know those well.

thinkyhead commented:
It seems like a straightforward enough option. We're just waiting for expert testimony from @teemuatlut —or others experienced with Trinamic drivers— whether it is considered prudent to expose this option to end-users.

mihtjel commented:
I did consider putting it in a section of its own, similarly to the CS pin/slave address/hybrid threshold settings - in that it's a thing for "if you know what you're doing". Though the worst consequence enabling it can have vs not is that you run _lower_ current to your motors, so it should not be destructive to enable by accident.

teemuatlut commented:
I don't think this should be merged in.
I think the negatives outweigh the positives as it adds unnecessary clutter to the config section and it's another thing a user can misconfigure. It advertises itself as an improvement and we still see users refer to rms settings as "vref" even when the analog input is disregarded. As a user configurable value it's not guaranteed to be accurate and so can largely negate the increased resolution. There are no boundary checks for dividing with 0 or using an otherwise invalid value like over 2.5. I am quite skeptical about any real tangible benefits there would be to having increased granularity. Typically the guideline for current setting is "enough but not too much".
I wasn't able to find the used formula used in this PR from the datasheet but it doesn't seem to take into account the changing V_fs setting that is set based on the configured Irms (=CS) value. On that note this also doesn't take into account the library already enables high sensitivity sense resistor, so the max rms value for low current applications is actually 0.98A and the resolution 31mA.
I appreciate the effort but I think there are better areas where we could use the dev time if you'd still be interested.

mihtjel commented:
Changing V_fs has no bearing on the scaling. It's from page 51 top of the tmc2209/8 datasheets, that happen to align at that point.

It would indeed fail (I think at compile time?) if any VREF is set to 0.

I offer no opinion on whether or not users are clever enough to understand anything, and should it not be merged, I'll probably keep a branch with it for as long as I need it myself.

mihtjel commented:
Apparently the compiler is kind, and will allow division by zero if you are really keen:

`Marlin\src\module\stepper\trinamic.cpp:48:37: warning: division by zero [-Wdiv-by-zero]
   #define EFFECTIVE_RSENSE(ST) (2.5 / ST##_VREF) * (ST##_RSENSE + 0.02) - 0.02`

thinkyhead commented:
> Apparently the compiler is kind, and will allow division by zero if you are really keen:

```cpp
#define EFFECTIVE_RSENSE(ST) (2.5 * RECIPROCAL(ST##_VREF)) * (ST##_RSENSE + 0.02) - 0.02
```

mihtjel committed:
Avoid divide by zero

mihtjel committed:
Move the configuration to a separate section
This should make it less intrusive for those not using it, and more
obviously an advanced feature that can be selected, but is not enabled
by accident.

mihtjel commented:
Thanks for the advice on avoiding the divide by zero, I made that change. I have also pushed a commit that moves the configuration to a separate section similarly to how CS/slave address/chopper settings/hybrid threshold is currently configured. I wonder if the chain_pos and interpolate settings might also improve clarity if they were moved in that manner.

thinkyhead commented:
This will not be merged until all issues are well understood and we can prevent settings that could cause damage.
